DenonAVR Config Flow

starkillerOG opened the pull request:
The denonavr library has been pushed to a new version 0.9.0, so this is now ready to be merged :)
Version 0.9.0 includes this PR wich is needed for Config Flow: https://github.com/scarface-4711/denonavr/pull/132

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
YAML configuration of the denonavr component has been deprecated, please remove the denonavr platform from your configuration.yaml file.
The denonavr component now uses the new Config Flow.
Denon and Marantz receivers should be discovered automatically and show up under "Configuration"-> "Integrations", from there hit configure and go through the steps, all previous setting options are still available during this setup.
If your receiver does not show up automatically, go to "Configuration"-> "Integrations"-> press the "+" icon -> search for "denonavr" and enter the setup.

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- Implement Config Flow
- Implement the new SSDP discovery using Config Flow
- Set device_ids and set unique ids for the receiver and there zones.
- Remove the depricated discovery implementation.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

YAML depricated --> Config Flow :)

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/35189 and https://github.com/home-assistant/core/issues/34871
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13319

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

starkillerOG committed:
DenonAVR Config Flow

starkillerOG committed:
DenonAVR Config Flow

starkillerOG committed:
fix spelling mistakes

starkillerOG committed:
Update .coveragerc

starkillerOG committed:
discovery DenonAVR

starkillerOG committed:
fix issort

bieniu commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,182 @@
+    @staticmethod
+    def construct_unique_id(model_name, serial_number):
+        """Construct the unique id from the ssdp discovery or user_step."""
+        return f"{model_name}-{serial_number}"
```
If I understand correctly, for a receiver that doesn't return `device_info` the `unique_id` will be `Unknown-<IP_ADDRESS>`. It seems to me not very unique.

starkillerOG replied:
I just commited a PR, improving the handeling of the unique_id and retrieving the serial_number.

bieniu commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -73,99 +54,30 @@
-    if receivers:
-        add_entities(receivers)
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Xiaomi Gateway Alarm from a config entry."""
```
`Xiaomi Gateway Alarm`?

starkillerOG replied:
sorry that was old code, I based this component on my previous PR of the Xiaomi Gateway Alarm.
just pushed a Commit to fix this

bieniu replied:
You don't have to be sorry. Copy/paste rulez ;)

starkillerOG committed:
improve serial_number handeling

starkillerOG committed:
fix old docstring

bieniu commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,199 @@
+            return self.async_abort(reason="connection_error")
+        receiver = connect_denonavr.receiver
+
+        mac_address = get_mac_address(ip=self.host)
```
If user will use hostname than IP address`get_mac_address` gives us `None`. You should use something like this https://github.com/home-assistant/core/blob/dev/homeassistant/components/huawei_lte/__init__.py#L332-L339

starkillerOG replied:
I don't think hostname will work with the denonavr library anyway, we schould use IP adress I think.

starkillerOG replied:
I will check if mac_address is None

starkillerOG replied:
fixed

bieniu replied:
I tested a component from your PR yesterday, I used the hostname and everything worked fine. So library works with hostnames.
Config_entry from the test:
```json
{
                "connection_class": "local_poll",
                "data": {
                    "host": "denon.bieniu.lan",
                    "receiver_id": "*AVR-X2100W-0005CD4556F4",
                    "show_all_sources": false,
                    "timeout": 2,
                    "zone2": false,
                    "zone3": false
                },
                "domain": "denonavr",
                "entry_id": "cbb47288fb934fdb9cda5a7cc16caad9",
                "options": {},
                "source": "user",
                "system_options": {
                    "disable_new_entities": false
                },
                "title": "Amplituner w salonie",
                "unique_id": "*AVR-X2100W-xxxxxxxx",
                "version": 1
            }
```

starkillerOG replied:
did not know that actually worked

starkillerOG replied:
I made another function to get the mac adress that now first tries to get the mac using ip, and otherwise use the hostname, so schould be fixed now

starkillerOG committed:
black styling

bieniu commented on homeassistant/components/denonavr/strings.json:
```diff
@@ -0,0 +1,40 @@
+        "title": "Denon AVR Network Receivers",
+        "description": "Connect to your receiver, if the IP adress is not set, auto-discovery is used",
+        "data": {
+          "host": "IP adress"
```
Should be `IP address`.

starkillerOG replied:
fixed

starkillerOG committed:
fix typo

starkillerOG committed:
improve getting mac

bieniu commented on homeassistant/components/denonavr/strings.json:
```diff
@@ -0,0 +1,40 @@
+    "step": {
+      "user": {
+        "title": "Denon AVR Network Receivers",
+        "description": "Connect to your receiver, if the IP adress is not set, auto-discovery is used",
```
Another `address` typo in the "description". Sorry, I haven't seen this before.

starkillerOG replied:
fixed

starkillerOG committed:
add mac to tests

starkillerOG committed:
fix get mac if hostname is used

bieniu commented:
The `aysnc_unload_entry` method is missing. Without this, the user will have to restart HA after removing the integration. Example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/brother/__init__.py#L50-L63

starkillerOG committed:
add async_unload_entry

starkillerOG committed:
fix unload entry

starkillerOG commented:
> 
> 
> The `aysnc_unload_entry` method is missing. Without this, the user will have to restart HA after removing the integration. Example: https://github.com/home-assistant/core/blob/dev/homeassistant/components/brother/__init__.py#L50-L63

Thanks, I have added the required function

starkillerOG committed:
black styling

starkillerOG commented:
@scarface-4711 Alright I think this PR is now completly ready, documentation update has been added and the tests have also been added and updated.
Now only the push of the denonavr library is needed

scarface-4711 commented:
`denonavr==0.9.0` which includes the changes is available now

scarface-4711 commented on homeassistant/components/denonavr/manifest.json:
```diff
@@ -1,7 +1,22 @@
   "documentation": "https://www.home-assistant.io/integrations/denonavr",
-  "requirements": ["denonavr==0.8.1"],
-  "codeowners": ["@scarface-4711", "@starkillerOG"]
+  "requirements": ["denonavr==0.8.1", "getmac==0.8.2"],
```
could be changed to `denonavr==0.9.0` now
same for requirements_all.txt and requirements_test_all.txt

starkillerOG replied:
done

starkillerOG committed:
denonavr version push 0.9.0

starkillerOG committed:
Merge branch 'dev' into patch-24

starkillerOG committed:
Improve setup when receiver has no power

starkillerOG committed:
Update tests

starkillerOG committed:
fix tests

starkillerOG commented:
@MartinHjelmare I think this PR is ready to be merged.

rroller commented on homeassistant/components/denonavr/strings.json:
```diff
@@ -0,0 +1,41 @@
+      },
+      "select": {
+        "title": "Select the receiver that you wish to connect",
+        "description": "Run the setup again if you want to connect aditional receivers",
```
`aditional` is misspelled. Should be `additional`

starkillerOG replied:
thanks, I just fixed it

rroller commented on homeassistant/components/denonavr/translations/en.json:
```diff
@@ -0,0 +1,41 @@
+      },
+      "select": {
+        "title": "Select the receiver that you wish to connect",
+        "description": "Run the setup again if you want to connect aditional receivers",
```
`aditional` spelled wrong here too

starkillerOG replied:
thanks, I just fixed it

rroller reviewed (commented):
Thanks for this!

starkillerOG committed:
update spelling + multiple receivers fix

starkillerOG committed:
remove * from model name

starkillerOG committed:
Update strings.json

starkillerOG committed:
Update en.json

starkillerOG committed:
Update config_flow.py

starkillerOG commented:
@MartinHjelmare can this still be included in the 0.110 release, or does this have to wait untill 0.111?
There are some people waiting on this to fix issue: https://github.com/home-assistant/core/issues/34871

MartinHjelmare commented:
This can't go into 0.110 as it contains new features and a breaking change.

starkillerOG commented:
Alright then aiming for 0.111.

pergolafabio commented:
What's so different? There are always breaking changed and new integration?

rroller commented:
Waiting for this to ship. Anyway to try it out earlier?

pergolafabio commented:
Yeah, can this be loaded as custom?

starkillerOG commented:
You can probably just download the changed files from this PR and use them to make a custom integration, but do'nt know how the "generated" files will work.

You can definitely just take these files and put them in your working installation folder and then you can use it.
However these files are overwritten if you update HomeAssistant + this might cause issues because this was build on top of dev so it will not work with HA 0.109 but it will work on top of 0.110. Be careful overwriting shared files like the generated, (only add the lines that were changed)

Fair warning: if you have no experience with the backend of HomeAssistant do not start messing with it, just wait till this gets merged. And always make a backup before you start messing :)

pergolafabio commented:
ill wait for 111 instead :)

starkillerOG commented:
@MartinHjelmare can you merge this?

scarface-4711 commented:
@starkillerOG you could update to `denonavr==0.9.1` before it gets merged

starkillerOG committed:
bump denonavr version to 0.9.2

starkillerOG commented:
@MartinHjelmare This PR has already been revieuwed by 3 people (not HomeAssistant maintainers) and has been basically stale and ready for merge for 14 days.

Could you please have a look and merge it if it is okay?

MartinHjelmare commented:
This is not a small PR. It will need to wait until someone has time to put in the time required for it.

We're currently swamped with PRs. We appreciate all the review help we can get.

starkillerOG commented:
@frenck this PR has been waiting on review and merge for 23 days already, could you please have a look at it if you have time?

bdraco commented on homeassistant/components/denonavr/__init__.py:
```diff
@@ -33,3 +43,54 @@ def service_handler(service):
+    hass: core.HomeAssistant, entry: config_entries.ConfigEntry
+):
+    """Set up the denonavr components from a config entry."""
+    if hass.data.get(DOMAIN) is None:
+        hass.data[DOMAIN] = {}
```
```suggestion
    hass.data.setdefault(DOMAIN, {})
```

bdraco commented on homeassistant/components/denonavr/__init__.py:
```diff
@@ -33,3 +43,54 @@ def service_handler(service):
+        entry.data[CONF_ZONE2],
+        entry.data[CONF_ZONE3],
+    )
+    await connect_denonavr.async_connect_receiver()
```
`raise ConfigEntryNotReady` if we cannot cannot so its retried later

starkillerOG replied:
thanks for the tip, done

bdraco commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        return f"{model_name}-{serial_number}"
+
+    async def async_get_mac(self, host):
+        """Get the mac address of the DenonAVR receiver."""
+        mac_address = await self.hass.async_add_executor_job(
+            partial(get_mac_address, **{"ip": host})
+        )
+        if not mac_address:
+            mac_address = await self.hass.async_add_executor_job(
+                partial(get_mac_address, **{"hostname": host})
+            )
```
`get_mac_address` can throw an exception here

starkillerOG replied:
What schould I do to resolve that?
Do you know what kind of exception, than I can put a try, except around it?

bdraco replied:
It cat at least throw `OSError` and probably a lot of others.  You'll probably need a broad check for `Exception` here.

starkillerOG replied:
alright I put a broad try except around it and made sure the mac_addres = None if we get an exception (and I log the exception), in that way we return self.async_abort(reason="no_mac").

Is this okay?

MartinHjelmare replied:
According to the docstring it will except exceptions silently and return `None`:
https://github.com/GhostofGoes/getmac/blob/3569b7ce503aa94f81a2d4cbab4a3aea8f44e39e/getmac/getmac.py#L119

starkillerOG replied:
So schould I then remove the try except again?

bdraco replied:
> So schould I then remove the try except again?

I've definitely that this throw on me before unexpectedly.  Let me take a look at the history to see if they fixed the issue recently as Martin's comment points out that they expect they have caught all exceptions possible.

bdraco replied:
It doesn't look like they have fixed the issue upstream.  I've found at least 3 places an exception could be thrown unexpectedly.

starkillerOG replied:
so put back the try catch?

bdraco replied:
I recall the issue was with `nmap_tracker`, but I had not realized at the time `getmac` was expected to trap exceptions as I would have saved the backtrace and already reported the issue upstream.

It needs to be fixed at some point since it is a bug in upstream code given the docstring.  Its likely going to be a rare case unless they have ipv6 disabled (then it will always happen) so there are two options here:

1. Put back in the try catch to insulate us from the unexpected behavior. The downside is this sweeps the problem under the rug but favors stability.

2. Let to blow up and report the issue upstream when/if we get an issue about it.  The downside here is the issue probably won't get fixed unless someone else reported it.  

It would be nice to get it fixed upstream, but I expect we need a backtrace for that to happen.  If you are going to be able to respond quickly to a report of an unexpected exception, report upstream, and get a fix out, go with option 2.  Otherwise go with option 1.

starkillerOG replied:
I no almost nothing about the whole getmac lib, so I will just use the try catch, better save than sorry :)

bdraco commented on homeassistant/components/denonavr/__init__.py:
```diff
@@ -33,3 +43,54 @@ def service_handler(service):
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        connections={(dr.CONNECTION_NETWORK_MAC, entry.data[CONF_MAC])},
+        identifiers={(DOMAIN, entry.unique_id)},
```
You only need one of these.  `connections` is sufficient.

starkillerOG replied:
I removed the identifiers

bdraco commented:
<img width="1053" alt="Screen Shot 2020-06-06 at 10 29 57 AM" src="https://user-images.githubusercontent.com/663432/83948114-b6d2c280-a7e0-11ea-94a9-ae8422d8c83a.png">

Manually adding works, but discovery didn't find it

bdraco commented:
[view-source_192.168.212.44_60006_upnp_desc_aios_device_aios_device.xml.zip](https://github.com/home-assistant/core/files/4740536/view-source_192.168.212.44_60006_upnp_desc_aios_device_aios_device.xml.zip)
Looks like the serialNumber is only in `urn:schemas-denon-com:device:AiosDevice:1`

```
<device>
  <deviceType>urn:schemas-denon-com:device:AiosDevice:1</deviceType>
  <friendlyName>Master Bedroom</friendlyName>
  <manufacturer>Marantz</manufacturer>
  <manufacturerURL>http://www.marantz.com</manufacturerURL>
  <modelName>Marantz NR1509</modelName>
  <modelNumber>Aios 4.025</modelNumber>
  <serialNumber>AZX36180400834</serialNumber>
  <UDN>uuid:64219b2d-e073-1232-0080-0005cdaee7c6</UDN>
  <DMH:X_WebAPIPort>8080</DMH:X_WebAPIPort>
<deviceList>
<device>
  <deviceType>urn:schemas-upnp-org:device:MediaRenderer:1</deviceType>
  <friendlyName>Master Bedroom</friendlyName>
  <manufacturer>Marantz</manufacturer>
  <manufacturerURL>http://www.marantz.com</manufacturerURL>
  <modelName>Marantz NR1509</modelName>
  <modelNumber>Aios 4.025</modelNumber>
  <UDN>uuid:889dd976-c74c-12c9-0080-0005cdaee7c6</UDN>
  <qq:X_QPlay_SoftwareCapability>QPlay:1</qq:X_QPlay_SoftwareCapability>
```

starkillerOG committed:
hass.data.setdefault

Co-authored-by: J. Nick Koston <nick@koston.org>

starkillerOG committed:
raise ConfigEntryNotReady if not connected

starkillerOG committed:
remove unnecesary identifiers

starkillerOG committed:
restore device identification

starkillerOG committed:
add zeroconf discovery for newer avr-x-2016 models

starkillerOG committed:
fix typo

starkillerOG commented:
@bdraco I fixed the discovery using "urn:schemas-upnp-org:device:MediaServer:1" instead of "urn:schemas-denon-com:device:AiosDevice:1" since that was also done like that in the deonavr library, see this PR: https://github.com/scarface-4711/denonavr/pull/136

Your receiver schould now be discovered using ssdp with this code

starkillerOG commented:
@bdraco are you able to merge this PR?

bdraco commented:
> @bdraco are you able to merge this PR?

Not yet, the getmac exception issue needs to be addressed

starkillerOG committed:
catch get_mac_address errors

starkillerOG committed:
disable=broad-except on get_mac_address

starkillerOG commented:
@bdraco what does this codecov/patch failed check mean?
How can I resolve that?

bdraco commented:
You have some missing test coverage in the config flow
`pytest --cov=homeassistant/components/denonavr/ --cov-report term-missing -- tests/components/denonavr/test_*.py`

bdraco commented:
`homeassistant/components/denonavr/config_flow.py     113      8    93%   146-153, 216-221`

Lines 146-153 and lines 216-221

starkillerOG commented:
Thanks!
I see, will resolve that tommorow

bdraco commented:
Also the latest update didn't cause discovery to show up

bdraco commented:
Here is what ssdp is discovering
```
2020-06-06 23:45:10 INFO (MainThread) [homeassistant.components.discovery] Unknown service discovered: dlna_dmr {'host': '192.168.212.44', 'port': 60006, 'ssdp_description': 'http://192.168.212.44:60006/upnp/desc/aios_device/aios_device.xml', 'name': 'Master Bedroom', 'model_name': 'Marantz NR1509', 'model_number': 'Aios 4.025', 'serial': 'AZX36180400834', 'manufacturer': 'Marantz', 'udn': 'uuid:64219b2d-e073-1232-0080-0005cdaee7c6', 'upnp_device_type': 'urn:schemas-denon-com:device:AiosDevice:1'}
```

bieniu commented on homeassistant/components/denonavr/__init__.py:
```diff
@@ -33,3 +44,53 @@ def service_handler(service):
+        connections={(dr.CONNECTION_NETWORK_MAC, entry.data[CONF_MAC])},
+        manufacturer=entry.data[CONF_MANUFACTURER],
+        name=entry.title,
+        model=f"{entry.data[CONF_MODEL]}-{entry.data[CONF_TYPE]}",
```
Why do we use the receiver type as part of the model string? For my AVR-2100W I get `AVR-X2100W-avr-x`.

starkillerOG replied:
Because that makes my job way easier ;)
There are 3 protocols used: avr, avr-x and avr-x-2016.
So if someone reports an issue, my first question will be to look in HomeAssistant and report the device model so that I now in which protocol the issue arises.

Unfortunately there is no separate property to specify protocol version in HomeAssistant

bieniu replied:
I'm not sure if we should change the way the device data is displayed to the user. Information about the type of device will be in the log.

starkillerOG committed:
add aditional ssdp discovery

starkillerOG committed:
revert try except for get_mac_address

starkillerOG committed:
bring config_flow tests to 100%

Also allow a failed serial_number retrival and using mac as backup

starkillerOG committed:
fix codespell

starkillerOG committed:
add get_mac_address try exception + test

starkillerOG committed:
add blank lines

bdraco commented:
Discovery is working 👍 

<img width="310" alt="Screen Shot 2020-06-07 at 9 43 36 AM" src="https://user-images.githubusercontent.com/663432/83971954-62444b80-a8a3-11ea-80b8-adcddaf415a3.png">
<img width="345" alt="Screen Shot 2020-06-07 at 9 43 11 AM" src="https://user-images.githubusercontent.com/663432/83971956-62dce200-a8a3-11ea-828e-87dbac7ebd63.png">

It would be nice if discovery showed the model/ip of the device that was discovered.   When there are multiple receivers.  `doorbird` has an example of how to do this.

starkillerOG commented:
I tryed  `self.context.update({"title_placeholders": {"name": "test"}})` but it did not work....
I think I will leave that up to a future PR.

starkillerOG commented:
@bdraco can you now merge this after the tests pass?

bdraco commented:
> I tryed `self.context.update({"title_placeholders": {"name": "test"}})` but it did not work....
> I think I will leave that up to a future PR.

Did you update `strings.json` as well?

starkillerOG commented:
nope that could be it, will try

bdraco commented:
> @bdraco can you now merge this after the tests pass?

Once we are though all the code comments, I'll need to do testing.  That will take a few hours.

starkillerOG committed:
add name to discovery

bdraco commented:
Docs look good.

starkillerOG commented:
@bdraco I just added the discovery name as you sugested

bdraco commented:
Discovery is showing it 👍 However after I add it instead of saying `Master Bedroom` is says `MAIN ZONE` for the config entry name.


<img width="306" alt="Screen Shot 2020-06-07 at 10 26 11 AM" src="https://user-images.githubusercontent.com/663432/83973047-64a9a400-a8a9-11ea-8c7e-04f238c1e176.png">
<img width="321" alt="Screen Shot 2020-06-07 at 10 26 01 AM" src="https://user-images.githubusercontent.com/663432/83973048-65423a80-a8a9-11ea-8cc0-9595372e0fe8.png">

starkillerOG commented:
@bdraco The fist name is the friendly name of the ssdp discovery (Master Bedroom)
The second name is the one that you get from the denonavr library, line 160:
`            title=receiver.name,`
So that schould be fixed in the denonavr lib

starkillerOG commented:
@bdraco the naming schould be fixed with this PR: https://github.com/scarface-4711/denonavr/pull/142

bdraco commented:
Sound mode reports Auto mode as Stereo 

Thats an existing issue though

starkillerOG commented:
> 
> 
> Sound mode reports Auto mode as Stereo
> 
> Thats an existing issue though

@bdraco what does `sound_mode_raw` attribute report if sound mode is Auto?

bdraco commented:
> > Sound mode reports Auto mode as Stereo
> > Thats an existing issue though
> 
> @bdraco what does `sound_mode_raw` attribute report if sound mode is Auto?


```
source_list:
  - AUX
  - Blu-ray/DVD
  - Bluetooth
  - CBL/SAT
  - CD
  - Game
  - HEOS Music
  - Media Player
  - Phono
  - TV Audio
  - Tuner
sound_mode_list:
  - MUSIC
  - MOVIE
  - GAME
  - AUTO
  - STANDARD
  - VIRTUAL
  - MATRIX
  - ROCK ARENA
  - JAZZ CLUB
  - VIDEO GAME
  - MONO MOVIE
  - DIRECT
  - PURE DIRECT
  - DOLBY DIGITAL
  - DTS SURROUND
  - MCH STEREO
  - STEREO
  - ALL ZONE STEREO
  - DOLBY AUDIO - DOLBY DIGITAL
volume_level: 0.575
is_volume_muted: false
media_content_type: channel
media_title: Blu-ray/DVD
source: Blu-ray/DVD
sound_mode: DOLBY PLII MOVIE
sound_mode_raw: Dolby PLII Movie
friendly_name: MAIN ZONE
supported_features: 69004
```

bdraco commented:
Without looking at the code  (since its out of scope for this pr). : `DOLBY PLII MOVIE` isn't a listed mode in the drop down so its probably defaulting to `STEREO`.

starkillerOG commented:
The state as above is working as expected, it is mapping "Dolby PLII Movie" to "DOLBY PLII MOVIE" which is correct.

sound_mode_raw is the sound_mode string response from the receiver, so if that does not show the correct sound mode there basically is nothing we can do, that is defined in the firmware of the device and can only be fixed by the manufacturer.

bdraco commented:
For clarity, `DOLBY PLII MOVIE` is not an option in the drop down

<img width="369" alt="Screen Shot 2020-06-07 at 11 44 49 AM" src="https://user-images.githubusercontent.com/663432/83974594-2e255680-a8b4-11ea-94eb-96216278938c.png">
<img width="398" alt="Screen Shot 2020-06-07 at 11 44 44 AM" src="https://user-images.githubusercontent.com/663432/83974595-2ebded00-a8b4-11ea-8f5d-5cfb6953618f.png">

starkillerOG commented:
@bdraco sorry I was mistaking, you were right.
I fixed that issue in this PR in the denonavr lib: https://github.com/scarface-4711/denonavr/pull/143

You schould have had a logger warning about this looking something like this:
`"Not able to match sound mode: 'DOLBY PLII MOVIE', returning raw sound mode."`

from line 1717 if denonavr:
```
_LOGGER.warning("Not able to match sound mode: '%s', "
                            "returning raw sound mode.", sound_mode_raw)
```

did you get that warning?

bdraco commented:
> did you get that warning?

Yes.  I didn't see it before because I was grepping for `denonavr` without `-i`

```
2020-06-07 17:26:06 INFO (SyncWorker_3) [DenonAVR] Identified receiver type: 'avr-x-2016' on port: '8080'
2020-06-07 17:26:06 INFO (MainThread) [homeassistant.setup] Setting up denonavr
2020-06-07 17:26:06 INFO (MainThread) [homeassistant.setup] Setup of domain denonavr took 0.0 seconds.
2020-06-07 17:26:06 INFO (SyncWorker_4) [DenonAVR] Identified receiver type: 'avr-x-2016' on port: '8080'
2020-06-07 17:26:06 INFO (MainThread) [homeassistant.components.media_player] Setting up media_player.denonavr
2020-06-07 17:26:06 WARNING (MainThread) [DenonAVR] Not able to match sound mode: 'Dolby Audio - Dolby Digital', returning raw sound mode.
2020-06-07 17:26:06 INFO (MainThread) [homeassistant.components.denonavr.media_player] Marantz receiver at host 192.168.212.44 initialized
2020-06-07 17:26:06 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new media_player.denonavr entity: media_player.main_zone
2020-06-07 18:29:57 ERROR (SyncWorker_13) [DenonAVR] Host 192.168.212.44 returned HTTP status code 403 to POST command at end point /NetAudio/index.put.asp
2020-06-07 18:30:00 ERROR (SyncWorker_9) [DenonAVR] Host 192.168.212.44 returned HTTP status code 403 to POST command at end point /NetAudio/index.put.asp
2020-06-07 18:30:02 ERROR (SyncWorker_13) [DenonAVR] Host 192.168.212.44 returned HTTP status code 403 to POST command at end point /NetAudio/index.put.asp
2020-06-07 18:30:46 WARNING (SyncWorker_13) [DenonAVR] Not able to match sound mode: 'Dolby PLII Movie', returning raw sound mode.
```

starkillerOG commented:
@bdraco I also added 'DOLBY AUDIO - DOLBY DIGITAL' to the sound mode mapping in the mentioned PR.
If you come accros more of those warnings, please report them in an issue in the denonavr lib so I can add those sound modes to the sound mode mapping

bdraco commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,234 @@
+        if user_input is not None:
+            # Get config option that have defaults
+            self.timeout = user_input[CONF_TIMEOUT]
+            self.show_all_sources = user_input[CONF_SHOW_ALL_SOURCES]
+            self.zone2 = user_input[CONF_ZONE2]
+            self.zone3 = user_input[CONF_ZONE3]
```
These should likely be moved to an options flow in a followup PR.

starkillerOG replied:
I was not aware of the existance of the options flow when I made this PR.
Will probably do that in the future, but first want this PR to be done.

scarface-4711 replied:
The config flow removes support for multiple zones in the current state. Several Denon receivers support multiple zones with different speakers set to different volumens and listening to different sources. In fact separate media player instances.
As far as I understand the concept of devices and entities, there should be 1 device for the Denon receiver with up to 3 media player entities for each enabled zone.

starkillerOG replied:
@scarface-4711 that is also the case, you still get 3 entities for the 3 diffrent zones if you check the zone2 and zone3 boxes

starkillerOG replied:
So it does not remove support for multiple zones, those are still supported

scarface-4711 replied:
I think you are right right 😄 
The debugger shows how the additional zones are queried. They probably do not appear because of the MAC adress issue, I just posted.

bdraco commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,234 @@
+        """Allow the user to specify settings."""
+        if user_input is not None:
+            # Get config option that have defaults
+            self.timeout = user_input[CONF_TIMEOUT]
```
Unless there is a strong case for adjustable timeouts, as developers we should pick a reasonable number here and not expose this as an option.  The user shouldn't have to know enough about how things work to make the decision.

starkillerOG replied:
well it was in the yaml configuration and I know there are some users that got problems with the default timeout of 2 seconds, espacially on old receiver models.
But for the majority 2 seconds is just fine, but it is nice to have it as an option if you are experiancing problems.

bdraco replied:
Is there any reason to not set it to 5 seconds?  That should cover all the use cases it seems.

starkillerOG replied:
I could set it to 5 seconds yes, that schould not be a problem I think.
Just do not want to upset users that are now able to configure the timeout and will not be if I remove it...

starkillerOG replied:
Do you want me to remove it?

bdraco replied:
Lets set it to 5 for now and take it out, if there is a use case presented that isn't covered by 5s, it can be added back to the options flow later.

starkillerOG replied:
I have removed the timeout option from the config flow and hard coded it to 5 seconds

bdraco commented:
Verified discovery via ssdp.

Verified discovery via not entering an ip.

Verified setup with an ip.

bdraco commented:
Verified removing the integration is successful

starkillerOG committed:
remove timeout option

remove the timeout option from the config flow and hardware code it to 5 seconds

starkillerOG commented:
@bdraco can this now be merged?

bdraco commented:
> @bdraco can this now be merged?

You seem very intent on merging this ASAP, is this change blocking other work?

bdraco commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -73,99 +56,38 @@
-    if receivers:
-        add_entities(receivers)
+        )
+    _LOGGER.info(
```
```suggestion
    _LOGGER.debug(
```

starkillerOG commented:
> 
> 
> > @bdraco can this now be merged?
> 
> You seem very intent on merging this ASAP, is this change blocking other work?

No not really blocking, but my PRs are stacking up and I just really want to wrap a few of them up and get them done.
This PR was intended as a quick implementation of the Config Flow for the denonavr platform since I already did some other Config flows. But it has been sitting here for a month now....

starkillerOG committed:
receiver initilized info->debug

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -237,6 +159,23 @@ def update(self):
+    def unique_id(self):
+        """Return the unique id of the zone."""
+        return self._unique_id
+
+    @property
+    def device_id(self):
+        """Return the device id of the receiver."""
+        return self._receiver_device_id
+
```
Is this still used?

starkillerOG replied:
nope will remove

bdraco commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -73,99 +56,38 @@
         self._name = self._receiver.name
+        self._unique_id = unique_id
+        self._mac = mac
+        self._receiver_device_id = receiver_device_id
```
Is this still used?

starkillerOG replied:
done

bdraco commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -73,99 +56,38 @@
     """Representation of a Denon Media Player Device."""
 
-    def __init__(self, receiver):
+    def __init__(self, receiver, receiver_device_id, unique_id, mac):
```
Here as well?

starkillerOG replied:
done

bdraco commented:
Testing looks good.  Only issue was the unmapped sound modes which is out of scope.

starkillerOG committed:
remove unused receiver_device_id

starkillerOG committed:
Merge branch 'patch-24' of https://github.com/starkillerOG/home-assistant into patch-24

bdraco commented:
Tested with a 120s sample with `py-spy`.  No noticeable performance impact or issues

bdraco commented:
Looks good. Just needs to be run though black and/or any other check format issues.

starkillerOG committed:
fix black styling

bdraco reviewed (approved):
Waiting for an approval from the existing code owner @scarface-4711 before merging.

scarface-4711 commented:
At the moment device integration does not work for me, because `get_mac_address` returns `None` which causes this exception.
Nevertheless, a success message appears on the UI, but there are neither devices nor entities listed. 

```
2020-06-07 23:53:35 INFO (SyncWorker_7) [DenonAVR] Identified receiver type: 'avr-x' on port: '80'
2020-06-07 23:53:35 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Denon AVR-X4100W for denonavr
Traceback (most recent call last):
  File "/workspaces/home-assistant/homeassistant/config_entries.py", line 215, in async_setup
    result = await component.async_setup_entry(  # type: ignore
  File "/workspaces/home-assistant/homeassistant/components/denonavr/__init__.py", line 72, in async_setup_entry
    device_registry.async_get_or_create(
  File "/workspaces/home-assistant/homeassistant/helpers/device_registry.py", line 131, in async_get_or_create
    connections = {
  File "/workspaces/home-assistant/homeassistant/helpers/device_registry.py", line 132, in <setcomp>
    (key, format_mac(value)) if key == CONNECTION_NETWORK_MAC else (key, value)
  File "/workspaces/home-assistant/homeassistant/helpers/device_registry.py", line 64, in format_mac
    if len(to_test) == 17 and to_test.count(":") == 5:
TypeError: object of type 'NoneType' has no len()
```

Apparently, it has to be ensured that `entry.data[CONF_MAC]` is not None at this place: **"/workspaces/home-assistant/homeassistant/components/denonavr/__init__.py", line 72**

Why is the mac address needed at all? I'm running home-assistant in the dev container setup. `getmac` seems to read `/proc/net/arp` in the running container, but there is no mac adress for my receiver listed. Instead I can find it in `/proc/net/arp` of the host machine. This probably won't happen only on my machine.

scarface-4711 commented:
Please also upgrade to `denonavr==0.9.3`

starkillerOG committed:
fix issue when mac unavailable

starkillerOG committed:
bump denonavr to 0.9.3

starkillerOG commented:
@scarface-4711 I have bumped the denonavr version to 0.9.3 and I schould have fixed the mac issue you were having.
Could you test and approve?

MartinHjelmare commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,239 @@
+            )
+            if not mac_address:
+                return self.async_abort(reason="no_mac")
+            self.serial_number = mac_address
```
We shouldn't set the mac address as the serial number and vice versa. Please keep these variables separate.

Introduce a third variable if needed to hold the unique_id that should be used for setting the unique_id.

starkillerOG replied:
I made a PR that prevents mixing of the serial_number as mac_address.

The other way around is more difficult since the mac_adress is also needed downsteam in the __init__.py file and again in the media_player.py file to idintify the enitity with the device.

Otherwise I would need to add an extra paramter to the Config entry, something like "device_identification" that will be identical to the mac_adress in normal cases, but will be diffrent if getting the mac failed.

However I think that is also messy since you then have another config flow data variable.
@MartinHjelmare Schould I add that extra var or leave it like this?

MartinHjelmare replied:
We should definitely not store the serial number as a mac address in the config entry data. Please change that. Name the data correctly and don't use the serial number for things that should use a mac address, like the connection set in device_info.

starkillerOG replied:
done no mixing anymore

starkillerOG committed:
do not mix serial and mac

starkillerOG committed:
remove dependancy on mac

MartinHjelmare commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,230 @@
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_settings(self, user_input=None):
```
Optional settings should not be part of the main flow but an options flow. If you don't want to add the options flow in this PR then remove these options for now from the flow.

starkillerOG replied:
I am not going to implement Options Flow. I have no idea how that works and frankly I am kind of done with this PR.
I can't remove those options because users will not be able to setup there zones and that will make a lot of users really angry...

starkillerOG replied:
done

MartinHjelmare commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -73,99 +54,32 @@
-NewHost = namedtuple("NewHost", ["host", "name"])
-
-
-def setup_platform(hass, config, add_entities, discovery_info=None):
```
If we would import the legacy yaml config to a config entry via the config flow we won't break existing user's options as much. Then we can wait with adding the options flow to the next PR without big impact.

starkillerOG replied:
I don't know how to mix the Config Flow with the old implementation and probably am going to run in a whole bunch of problems with that. Not going to open that can of worms...
I do understand that this will be a breaking change but that is the case with many of the already accepted Config Flow implementations.

starkillerOG replied:
not needed anymore

starkillerOG committed:
fix styling and tests

starkillerOG commented:
I have put a lot of effort in this PR but at this point I am not willing to put much more into it.
It litraly takes me a few days (like 3 days) to implement this PR and make it working which does not take much effort.
Than it takes me more than a month to get it accepted and merged and I am not even done with this PR, that kind of takes the fun out of developing...

At this point there are 3 options:
1) you think that the current state of this PR is going to improve the current implemetation --> than this schould be merged.
2) you think that the current state of this PR is going to make the current implementation worse --> than just close this PR.
3) you think this cannot be merged withouth big changes but don't want to close it --> than someone else will need to be found that is willing to make those changes. I will accept PR's on the patch-24 branch to improve this PR.

I will make sure that all test pass tho.
I hope you will consider merging this PR.

MartinHjelmare commented:
We'll not accept PRs that are not following our standards. If you're not going to continue here, please close the PR.

MartinHjelmare reviewed (changes requested):
Requesting changes to reflect the correct status of this PR. See above.

starkillerOG committed:
Move settings to options flow

starkillerOG commented:
@MartinHjelmare I added the options flow as you insisted.
can this now be approved?

bdraco commented:
Thanks for coming back to this 👍   I'll run though this in a bit. 

I noticed the update listener isn't unsubscribed which will cause a duplicate listener after the config entry is reloaded.

There is an example in `nut` 
```
undo_listener = entry.add_update_listener(_async_update_listener)
```

bdraco commented on homeassistant/components/denonavr/__init__.py:
```diff
@@ -33,3 +48,70 @@ def service_handler(service):
+
+    device_registry = await dr.async_get_registry(hass)
+    if entry.data[CONF_MAC] is not None:
+        device_registry.async_get_or_create(
```
You could save a few lines with something like this

```
       device_info = {
           "identifiers": ...
       }
       if entry.data[CONF_MAC] is not None:
           device_info["connections"] = {(dr.CONNECTION_NETWORK_MAC, entry.data[CONF_MAC])}

        device_registry.async_get_or_create(**device_info)
```

bdraco replied:
Actually it probably makes this a lot simpler if all of this is moved into `def device_info` in `media_player.py`

starkillerOG replied:
If I move it to the device itself, I would have to pass all the config_entry data required for it, so I would prefer to keep it in the __init__.py file

bdraco replied:
From my read of the documentation, it implies that `async_get_or_create` is for when there is no entity so I think it should be moved into `device_info`

We could wait for Martin to give a second opinion though if you don't concur.

starkillerOG replied:
Lets wait on the revieuw of @MartinHjelmare to see what he thinks. I don't have strong feelings on it. I'm fine with moving it to the entity, but personly i think it is cleaner as it is now. So I will let @MartinHjelmare decide...

MartinHjelmare replied:
We should only call `device_registry.async_get_or_create` when we need to create a device that is not the same device as the device that has a feature represented by an entity. Eg if there's a hub that connects multiple other devices, we can create a home assistant device to represent the hub, while the other devices will be represented by entities which will automatically create home assistant devices via the `device_info` property.

starkillerOG replied:
I will move this to the entity then, will work on it this evening

starkillerOG replied:
Done and tested

starkillerOG committed:
fix undo update listener

starkillerOG committed:
reduce duplication of device_info

bdraco commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -237,6 +150,18 @@ def update(self):
+        """Return the unique id of the zone."""
+        return self._unique_id
+
+    @property
+    def device_info(self):
+        """Return the device info of the receiver."""
+        return {
+            "identifiers": {(DOMAIN, self._receiver_device_id)},
+        }
+
```
```suggestion
```

Since you do this in `__init__.py` this can come out

bdraco replied:
Also `self._receiver_device_id` , and it being passed into create the object.

bdraco commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -237,6 +150,18 @@ def update(self):
+        """Return the unique id of the zone."""
+        return self._unique_id
+
+    @property
+    def device_info(self):
+        """Return the device info of the receiver."""
+        return {
+            "identifiers": {(DOMAIN, self._receiver_device_id)},
+        }
+
```
I was hoping this could be removed, but there is no way to link the entity to the device without it

starkillerOG replied:
Yes that is needed to link the device to the entity indeed

MartinHjelmare commented on homeassistant/components/denonavr/strings.json:
```diff
@@ -37,5 +32,18 @@
+    "step": {
+      "init": {
+        "title": "Denon AVR Network Receivers",
+        "description": "Specify optional settings, a restart is required to apply the settings",
```
We have the config entry listener to reload the entry, so a restart isn't needed, right?

starkillerOG replied:
Yes I will remove that, that was before I used the config entry listener.

One question about this: when I add an extra zone everything works fine and the extra entity is created. However if I delete a zone, the entity of that zone will show as unavailable but still be there. If I then restart HomeAssistant it will show as an orphan entity (or wheter it is called with one of those red exclamation marks "!") and I then need to hit the "remove entity" to get rid of it.

Am I doing something wrong here?
Is there a better way to completly get rid of the entity inside the config entry listener besides just reloading the integration and not setting up the entity anymore?

MartinHjelmare replied:
If we change what entities the integration creates we also need to clean up the registries, entity and device registries. We have these methods available:

https://github.com/home-assistant/core/blob/3cc94f7d6a0364e7347bc526d5e9251f8f2b8169/homeassistant/helpers/entity_registry.py#L285

https://github.com/home-assistant/core/blob/3cc94f7d6a0364e7347bc526d5e9251f8f2b8169/homeassistant/helpers/device_registry.py#L322

starkillerOG replied:
@MartinHjelmare thanks!
The device_registry will not be needed since there is only one device that is always there, only the 2nd and 3th zones entities can be added/removed but they belong to the same device as the main zone entity which is always there.

I will add the async_remove function this evening.

Is there an easy way to check which entity is beeing removed?
possibilities:
 - compare the old and new config_entry.options, but how do I get the old config_entry.options
 - check if the zone 2 and 3 entities exist and compare to the new config_entry.options, but how to properly check if an entitie exists?

(I can probably make it work, but that will probably not be the best way to do it)

MartinHjelmare replied:
We can get the existing entity registry entries for a device:
https://github.com/home-assistant/core/blob/1b6f0b78aa7b70a70d16de4b0a22b302b7f4b0d9/homeassistant/helpers/entity_registry.py#L507-L510

Then remove all the entries that doesn't match with the new zones.

starkillerOG replied:
Nice, thanks!
will do that

starkillerOG replied:
Done and tested to work, thanks for your help!

MartinHjelmare commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,255 @@
+        else:
+            _LOGGER.error(
+                "Could not get serial number of host %s, "
+                "using the mac_address as identification",
```
Can we assume that a device will use the same unique id, serial or mac, regardless if it's set up via the user or via discovery?

What devices do not provide a serial number? How often does that happen?

starkillerOG replied:
Yes it will in principle always use the serial number regardless of the way you set up (manually or discovery).

If the serial is not available on discovery the manually setup will also not have the serial number available because the denonavr lib uses the same xml file to get the serial as the discovery component does.

However we can of course not 100% garantuee that this is the case, but the whole using mac_address is just a backup method that issues errors so that users can report the error but in the mean time be able to use the integration. We schould use the serial number and if that is not available users schould report this error and it schould be fixed.

MartinHjelmare replied:
Mac address is also not always available so I don't think it's a good fall back.

I think it would be better to fall back to no unique id, when serial number is missing, and then just check the existing config entry host addresses to make sure we don't set up the same host again. With no unique id, we return `None` for `device_info` and `unique_id` in the entity.

starkillerOG replied:
All devices that are tested and known to me always provide serial number.
However since Denon/Marantz have the really enoying habit of implementing there API just a litlle bit diffrent for each of there models I can not 100% garantuee that it will work for all models.

(like sound modes are reported with extra spaces or just spelled diffrently in the denon/marantz API, really anoying... for example 'MULTI CH STEREO', 'MULTI_CH_STEREO', 'MCH STEREO')

MartinHjelmare replied:
Yeah, I think that's another reason why we shouldn't make this too complicated. If the known devices report serial number, let's go with that and fall back to no unique id otherwise.

starkillerOG replied:
alright, will change that tonight

starkillerOG replied:
I removed the mac_adress fallback and allowed the integration to setup withouth an unique_id.

starkillerOG committed:
Properly remove zone2 and zone3

starkillerOG committed:
move device_info to entity

starkillerOG committed:
remove mac_adress fallback for unique_id

starkillerOG committed:
remove unused string

starkillerOG committed:
properly handle no unique_id

starkillerOG committed:
fix tests

starkillerOG commented:
@MartinHjelmare I have processed all your feedback.
Can this now be approved?

MartinHjelmare commented on homeassistant/components/denonavr/__init__.py:
```diff
@@ -98,21 +106,36 @@ async def async_setup_entry(
+    for entry in entries:
+        if entry.unique_id == zone2_id and not config_entry.options.get(CONF_ZONE2):
+            entity_registry.async_remove(entry.entity_id)
+            _LOGGER.info("removing zone2 from DenonAvr")
```
```suggestion
            _LOGGER.debug("Removing zone2 from DenonAvr")
```

MartinHjelmare commented on homeassistant/components/denonavr/__init__.py:
```diff
@@ -98,21 +106,36 @@ async def async_setup_entry(
+            _LOGGER.info("removing zone2 from DenonAvr")
+        if entry.unique_id == zone3_id and not config_entry.options.get(CONF_ZONE3):
+            entity_registry.async_remove(entry.entity_id)
+            _LOGGER.info("removing zone3 from DenonAvr")
```
```suggestion
            _LOGGER.debug("Removing zone3 from DenonAvr")
```

MartinHjelmare commented on homeassistant/components/denonavr/media_player.py:
```diff
@@ -158,9 +159,19 @@ def unique_id(self):
-        return {
-            "identifiers": {(DOMAIN, self._receiver_device_id)},
+        device_info = {
+            "config_entry_id": self._config_entry.entry_id,
```
We don't need to set this. It's set by the core.

starkillerOG replied:
I have removed it

MartinHjelmare commented on homeassistant/components/denonavr/config_flow.py:
```diff
@@ -0,0 +1,253 @@
+                return await self.async_step_connect()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
```
We need to guard for setting up the same host more than once, since we don't guarantee a unique id.

starkillerOG replied:
How do you propose to do that?
I thought the unique_id was speciffically ment for that, that was the reason I aborted setup if unique_id could not be set and had a fallback for the unique_id...

starkillerOG replied:
I could just abort the setup if no serial number could be found?

MartinHjelmare replied:
Yes, the config entry unique_id is used to guard from setting up the same id more than once, but since we can't guarantee that a unique id is available we need another way too that guards for this.

Something like this:
```py
for entry in self._async_current_entries():
    if entry.data[CONF_HOST] == host:
        return self.async_abort(reason="already_configured")
```

starkillerOG replied:
I added it, and also made a test for it.
I used entry.data.get(CONF_HOST) because I think the self._async_current_entries() can also give entities of other integrations right? That do not nessesarly have the CONF_HOST?

MartinHjelmare replied:
No, it will only return entries for this domain. Otherwise it would not work for our purposes.

starkillerOG replied:
alright will make it a key acces then

starkillerOG replied:
done

starkillerOG committed:
reduce log level to debug

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
reduce log level to debug

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

starkillerOG committed:
remove config_entry_id

starkillerOG committed:
check for double configuration when no serial + test

starkillerOG committed:
Merge branch 'patch-24' of https://github.com/starkillerOG/home-assistant into patch-24

MartinHjelmare commented on tests/components/denonavr/test_config_flow.py:
```diff
@@ -188,7 +188,7 @@ async def test_config_flow_manual_discover_error(hass):
 
 async def test_config_flow_manual_host_no_serial(hass):
-    """Test a successful config flow manually initialized by the user with the host specified and an error getting the serial number but backup mac address success."""
+    """Test a successful config flow manually initialized by the user with the host specified and an error getting the serial number."""
```
Please break long strings. The docstring should start with a header line that fits on a single line, after that should optionally follow a blank line and then a body. So write a shorter header line and follow up with a body if needed.

PEP8 says that docstrings should be limited to 72 characters width:
https://www.python.org/dev/peps/pep-0008/#maximum-line-length

starkillerOG replied:
I updated the docstrings

starkillerOG committed:
Update docstrings

starkillerOG commented:
@MartinHjelmare any more feedback?

MartinHjelmare reviewed (approved):
Looks good!

scarface-4711 commented:
Good work @starkillerOG 😄

rroller commented:
Thank you! I’ve been waiting for this
